lib/Common/DataStructures/BaseDictionary.h

```diff
@@ -860,9 +860,14 @@ namespace JsUtil
             int initSize = max(capacity, 4);
             uint initBucketCount = SizePolicy::GetBucketSize(initSize);
             AssertMsg(initBucketCount > 0, "Size returned by policy should be greater than 0");
-            Allocate(&buckets, &entries, initBucketCount, initSize);
 
-            // Allocation can throw - assign the size only after allocation has succeeded.
+            int* newBuckets = nullptr;
+            EntryType* newEntries = nullptr;
+            Allocate(&newBuckets, &newEntries, initBucketCount, initSize);
+
+            // Allocation can throw - assign only after allocation has succeeded.
+            SetArray<TAllocator>(this->buckets, newBuckets, initBucketCount);
```

Comment:
The write barrier on elements scenario is hard to understand (I'm still not too clear after asking Lei and Curtis.) `this->buckets = newBuckets` is fine here, but `this->entries = newEntries` is not. I was thinking of how we prevent accidental error. In the case of pointer field, the `operator=` function takes care of all scenarios, be it raw pointer, NoWriteBarrierPtr, or WriteBarrierPtr. So in comparison, I was thinking maybe we require all array assignment through this `SetArray` function. `SetArray` determines if we need write barrier or not. Possibly we can track that in clang plugin (not trivial...).